pkg/operatorsource/configuring_test.go

```diff
@@ -9,12 +9,110 @@ import (
+	fakeclient := NewFakeClient()
+	refresher := mocks.NewSyncerPackageRefreshNotificationSender(controller)
+
+	reconciler := operatorsource.NewConfiguringReconcilerWithInterfaceClient(helperGetContextLogger(), factory, writer, fakeclient, refresher)
+
+	ctx := context.TODO()
+	opsrcIn := helperNewOperatorSourceWithPhase("marketplace", "foo", phase.Configuring)
+
+	opsrcWant := opsrcIn.DeepCopy()
+	opsrcWant.Status.Packages = "etcd,prometheus,amqp"
+
+	registryClient := mocks.NewAppRegistryClient(controller)
+
+	optionsWant := appregistry.Options{Source: opsrcIn.Spec.Endpoint}
+	factory.EXPECT().New(optionsWant).Return(registryClient, nil).Times(1)
+
+	// We expect the remote registry to return a non-empty list of manifest(s).
+	manifestExpected := []*datastore.RegistryMetadata{
+		&datastore.RegistryMetadata{
+			Namespace:  "redhat",
+			Repository: "myapp",
+			Release:    "1.0.0",
+			Digest:     "abcdefgh",
+		},
+	}
+	registryClient.EXPECT().ListPackages(opsrcIn.Spec.RegistryNamespace).Return(manifestExpected, nil).Times(1)
+
+	// We expect the datastore to save downloaded manifest(s) returned by the registry.
+	writer.EXPECT().Write(opsrcIn, manifestExpected).Return(1, nil)
+
+	// The first time we ask for the packages from the datastore, we expect to get nothing.
+	writer.EXPECT().GetPackageIDsByOperatorSource(opsrcIn.GetUID()).Return("")
+
+	// Then we expect to call send refresh, because the package list was empty.
+	refresher.EXPECT().SendRefresh()
+
+	// We expect datastore to return the specified list of packages.
+	writer.EXPECT().GetPackageIDsByOperatorSource(opsrcIn.GetUID()).Return(opsrcWant.Status.Packages)
+
+	opsrcGot, nextPhaseGot, errGot := reconciler.Reconcile(ctx, opsrcIn)
+
+	assert.NoError(t, errGot)
+	assert.Equal(t, opsrcWant, opsrcGot)
+	assert.Equal(t, nextPhaseWant, nextPhaseGot)
+}
+
+// Use Case: Registry returns an empty list of metadata.
+// Expected Result: Next phase is set to "Failed".
+func TestReconcile_OperatorSourceReturnsEmptyManifestList_Failed(t *testing.T) {
+	controller := gomock.NewController(t)
+	defer controller.Finish()
+
+	writer := mocks.NewDatastoreWriter(controller)
+	factory := mocks.NewAppRegistryClientFactory(controller)
+	fakeclient := NewFakeClient()
+	refresher := mocks.NewSyncerPackageRefreshNotificationSender(controller)
+
+	reconciler := operatorsource.NewConfiguringReconciler(helperGetContextLogger(), factory, writer, fakeclient, refresher)
+
+	ctx := context.TODO()
+	opsrcIn := helperNewOperatorSourceWithPhase("marketplace", "foo", phase.Configuring)
+
+	registryClient := mocks.NewAppRegistryClient(controller)
+
+	optionsWant := appregistry.Options{Source: opsrcIn.Spec.Endpoint}
+	factory.EXPECT().New(optionsWant).Return(registryClient, nil).Times(1)
+
+	// We expect the registry to return an empty manifest list.
+	manifests := []*datastore.RegistryMetadata{}
+	registryClient.EXPECT().ListPackages(opsrcIn.Spec.RegistryNamespace).Return(manifests, nil).Times(1)
```

Comment:
I don't think the `Times(1)` is not needed. By default `EXPECT()` expects a single call to that function.

Reply:
I didn't write this as part of this PR. Not sure this is a refactor that should go in here

Reply:
Interesting, All of these test cases are being shown as additions in this PR. Maybe @tkashem added the `Times(1)` and can help us understand if my understanding is correct about it.

Reply:
These test cases were moved from `downloading_test.go`